types/src/key.rs

```diff
@@ -239,39 +234,73 @@ impl Key {
-            Ok(Key::Account(account_hash))
-        } else if let Some(hex) = input.strip_prefix(HASH_PREFIX) {
-            Ok(Key::Hash(HashAddr::try_from(
-                base16::decode(hex)?.as_ref(),
-            )?))
-        } else if let Some(hex) = input.strip_prefix(DEPLOY_INFO_PREFIX) {
-            Ok(Key::DeployInfo(DeployHash::new(
-                <[u8; DEPLOY_HASH_LENGTH]>::try_from(base16::decode(hex)?.as_ref())?,
-            )))
-        } else if let Ok(transfer_addr) = TransferAddr::from_formatted_str(input) {
-            Ok(Key::Transfer(transfer_addr))
-        } else if let Ok(uref) = URef::from_formatted_str(input) {
-            Ok(Key::URef(uref))
-        } else if let Some(era_id_str) = input.strip_prefix(ERA_INFO_PREFIX) {
-            Ok(Key::EraInfo(EraId::from_str(era_id_str)?))
-        } else if let Some(hex) = input.strip_prefix(BALANCE_PREFIX) {
-            Ok(Key::Balance(URefAddr::try_from(
-                base16::decode(hex)?.as_ref(),
-            )?))
-        } else if let Some(hex) = input.strip_prefix(BID_PREFIX) {
-            Ok(Key::Bid(AccountHash::new(AccountHashBytes::try_from(
-                base16::decode(hex)?.as_ref(),
-            )?)))
-        } else if let Some(hex) = input.strip_prefix(WITHDRAW_PREFIX) {
-            Ok(Key::Withdraw(AccountHash::new(AccountHashBytes::try_from(
-                base16::decode(hex)?.as_ref(),
-            )?)))
-        } else if let Some(era_id_str) = input.strip_prefix(VALIDATORS_PREFIX) {
-            Ok(Key::EraValidators(EraId::from_str(era_id_str)?))
-        } else {
-            Err(FromStrError::InvalidPrefix)
+        if input.starts_with(ACCOUNT_HASH_FORMATTED_STRING_PREFIX) {
+            let account_hash = AccountHash::from_formatted_str(input)?;
+            return Ok(Key::Account(account_hash));
+        }
+
+        if let Some(hex) = input.strip_prefix(HASH_PREFIX) {
+            let addr =
+                base16::decode(hex).map_err(|error| FromStrError::Hash(error.to_string()))?;
+            let hash_addr = HashAddr::try_from(addr.as_ref())
+                .map_err(|error| FromStrError::Hash(error.to_string()))?;
+            return Ok(Key::Hash(hash_addr));
+        }
+
+        if let Some(hex) = input.strip_prefix(DEPLOY_INFO_PREFIX) {
+            let hash =
+                base16::decode(hex).map_err(|error| FromStrError::DeployInfo(error.to_string()))?;
+            let hash_array = <[u8; DEPLOY_HASH_LENGTH]>::try_from(hash.as_ref())
+                .map_err(|error| FromStrError::DeployInfo(error.to_string()))?;
+            return Ok(Key::DeployInfo(DeployHash::new(hash_array)));
         }
+
+        if input.starts_with(TRANSFER_ADDR_FORMATTED_STRING_PREFIX) {
+            let transfer_addr = TransferAddr::from_formatted_str(input)?;
+            return Ok(Key::Transfer(transfer_addr));
+        }
+
+        if input.starts_with(UREF_FORMATTED_STRING_PREFIX) {
```

Comment:
Not a blocker but just something that occurred to me that there is duplicated logic - first, this checks prefix with `starts_with` and then `URef::from_formatted_str` does `strip_prefix` inside doing the same work twice. Alternatively, wouldn't this work for `T::from_formatted_str` if we already have T's error type connected through From traits?

```rust
match URef::from_formatted_str(input) {
  Ok(uref) => return Key::URef(uref),
  Err(uref::FromStrError::InvalidPrefix) => {}
  Err(e) => return Err(e.into()),
}
```

Reply:
Yup, makes sense.  Updated in [28599eb](https://github.com/CasperLabs/casper-node/pull/1231/commits/28599ebdd84231ce637399d24bbafb75e900a2a3).